microraft-store-sqlite/src/main/java/io/microraft/store/sqlite/RaftSqliteStore.java

```diff
@@ -0,0 +1,346 @@
+    }
+
+    @Override
+    public void truncateLogEntriesFrom(long logIndexInclusive) {
+        dsl.deleteFrom(LOG_ENTRIES)
+                .where(INDEX.greaterOrEqual(logIndexInclusive))
+                .execute();
+    }
+
+    @SuppressWarnings("VarUsage")
+    private Select<? extends Record1<Long>> completedSnapshots() {
+        Table<?> tempTable = DSL.table("t");
+        var blocks = dsl.select(
+                        qualify(SNAPSHOT_CHUNKS, INDEX),
+                        DSL.count(qualify(SNAPSHOT_CHUNKS, INDEX)).as(COUNT))
+                .from(SNAPSHOT_CHUNKS)
+                .groupBy(qualify(SNAPSHOT_CHUNKS, INDEX));
+        return dsl.select(qualify(tempTable, INDEX))
+                .from(blocks.asTable(tempTable), SNAPSHOTS)
+                .where(qualify(tempTable, INDEX).eq(qualify(SNAPSHOTS, INDEX)))
+                .and(DSL.field(COUNT).eq(CHUNK_COUNT));
+    }
+
+    @Override
+    public void truncateSnapshotChunksUntil(long logIndexInclusive) {
+        dsl.deleteFrom(SNAPSHOT_CHUNKS)
+                .where(qualify(SNAPSHOT_CHUNKS, INDEX).le(logIndexInclusive))
+                .and(qualify(SNAPSHOT_CHUNKS, INDEX).notIn(completedSnapshots()))
+                .execute();
+    }
+
+    // Visible for testing
+    List<SnapshotChunk> getAllSnapshotChunks() {
+        return dsl.select(chunkField)
+                .from(SNAPSHOT_CHUNKS)
+                .orderBy(INDEX, CHUNK_INDEX)
+                .fetch(chunkField);
+    }
+
+    @Override
+    public void flush() {
+        dsl.connection(Connection::commit);
+        if (tryToCleanUpOldSnapshots) {
+            tryToCleanUpOldSnapshots = false;
+            Optional<Long> maybeSnapshotIndex = getMaxCommittedSnapshotIndex();
+            maybeSnapshotIndex.ifPresent(snapshotIndex -> {
+                dsl.deleteFrom(LOG_ENTRIES)
+                        .where(INDEX.lessOrEqual(snapshotIndex))
+                        .execute();
+                dsl.deleteFrom(SNAPSHOT_CHUNKS)
+                        .where(INDEX.lessThan(snapshotIndex))
+                        .execute();
+                dsl.deleteFrom(SNAPSHOTS).where(INDEX.lessThan(snapshotIndex)).execute();
+                dsl.connection(Connection::commit);
+            });
+        }
+    }
+
+    public Optional<RestoredRaftState> getRestoredRaftState() {
+        var record = dsl.select(
+                        localEndpointField, LOCAL_ENDPOINT_VOTING, initialGroupMembersField, TERM, votedForField)
+                .from(KV)
+                .fetchOne();
+
+        if (record == null
+                || record.get(localEndpointField) == null
+                || record.get(LOCAL_ENDPOINT_VOTING) == null
+                || record.get(initialGroupMembersField) == null
+                || record.get(TERM) == null
+                || record.get(votedForField) == null) {
+            return Optional.empty();
+        }
```

Comment:
should we differentiate the 2 cases here:
- if record is null, then this is an empty state being initialized for the first time. then we can return empty optional.
- else, if any of localEndpointField, LOCAL_ENDPOINT_VOTING, initialGroupMembersField is null, or if voted for is non-null but term is null, then we should throw because this is a broken state.

with this, we prevent the case where a Raft node starts with an empty state mistakenly while its persisted state gets corrupted.

Reply:
I need more advice on this one, I think. I don't think the cases you described are quite the right ones; you can see in createTables that we create the row every time (the row always exists).

I've added some assertions that I think are as strong as they can be without introducing the potential for false positives.